cli/internal/turbopath/turbopath.go

```diff
@@ -0,0 +1,180 @@
+// Package turbopath teaches the Go type system about six
+// different types of paths:
+// - AbsoluteSystemPath
+// - RelativeSystemPath
+// - RepoRelativeSystemPath
+// - AbsoluteUnixPath
+// - RelativeUnixPath
+// - RepoRelativeUnixPath
+//
+// There is a second portion which allows clustering particular
+// dimensions of these four types:
+// - AbsolutePathInterface
+// - RelativePathInterface
+// - RepoRelativePathInterface
+// - UnixPathInterface
+// - SystemPathInterface
+// - FilePathInterface
+//
+// Between these two things it is assumed that we will be able to
+// reasonably describe file paths being used within the system and
+// have the type system enforce correctness instead of relying upon
+// runtime code to accomplish the task.
+//
+// Much of this is dreadfully repetitive because of intentional
+// limitations in the Go type system.
+package turbopath
```

Comment:
I got curious and started deleting things from this package to see what I could get away with. I think some of our type system woes can potentially be addressed by just not defining some things. For instance, it appears we never call `RepoRelativeUnixPath.ToRepoRelativeSystemPath`. That, combined with other things like that, mean that I don't think we need `RepoRelativeSystemPath`. 

I'll put up a PR to show what I mean, but I think by dropping some thing from interfaces when we aren't actually using the interface methods, we can return more useful types.

Reply:
Just to show what I'm talking about: https://github.com/nathanhammond/turborepo/pull/2

Reply:
So I did actively choose to implement the full set because my intent was to treat this as a library—and possibly one which could be fully-externalized (move out of `internal`, and then maybe even into a separate repo). That would also make it a bit easier to optionally include it since it *does* have runtime costs. In some (far-)future opportunity we could even add a cute little AST transform that enables it to remove itself from runtime code.

The tradeoff here is that if we optimize it for the paths that we have right now it results in an inconsistent interface and stops feeling like a library and starts to feel more like a whole bunch of utility functions.

I do understand the appeal of reducing the use of interfaces: using them can add to verbosity and generally feels unergonomic. But I think it's worth it in this case.

Also, I'm ~sure that this is the full set of things will eventually be used. Repo-relative system paths (the one explicitly added to mirror repo-relative Unix paths which is ~`00% unused) are going to be useful if we add rewrite-symlink-on-cache for Windows.

Reply:
I further refactored application code more to enable removing interfaces:
https://github.com/vercel/turborepo/pull/1239/commits/ae13799cc89d2090827b94a896f375da39554f67

I continued on to remove all Interfaces from `turbopath`:
https://github.com/vercel/turborepo/pull/1239/commits/71b515e2a11ebf185eeccd1755942624d61b724b

But the path to get to where we are goes through the interfaces that I removed, for example:
https://github.com/vercel/turborepo/pull/1239/commits/c27caff8073868eee9e62a5fce2784badf8de098

I'm *very much* not a fan of losing the tool I used to get where this PR is. `package_deps_hash.go` is nearly a full rewrite and removing the interfaces makes traveling that path impossible. It's all showing up as a single PR at the end, but the incremental changes to get there are captured in the history, go through the interfaces, and are absolutely necessary when adopting a complicated mess of unknown paths into the system.

Having gone through this process I feel *very* strongly that removing the interfaces is a bad idea. If I were to do the same thing for, say, caching, my *first* step would be to reintroduce them as guiderails for the effort.

Reply:
I spent some time reading through these.

Maybe we need to chat offline about it, because I'm not 100% sure I'm understanding the issue. For instance, going from [`c27caff`](https://github.com/vercel/turborepo/commit/c27caff8073868eee9e62a5fce2784badf8de098) to the current implementation looks like a clear improvement to me. To pick one example, having `filesToHash` as an interface argument to `gitHashObject` suggests that we don't know what kind of paths we're dealing with at that point. That stands out as the sort of thing these types are intended to fix: we _should_ know exactly what kind of path we have there. And in what we have now, that is in fact the case, we know they are `AnchoredSystemPath` instances.

[`ae13799`](https://github.com/vercel/turborepo/commit/ae13799cc89d2090827b94a896f375da39554f67) also looks like a clear win to me: aside from the `Relative` -> `Anchored` change, the only change outside of the `turbopath` package, I think, is changing from an `AbsolutePathInterface` to `AbsoluteSystemPath`. That's doing what I think we want: giving us more explicit information about exactly what kind of path we have.

I also have no objection to using interfaces as a tool for getting to an implementation using explicit types, or in cases where we do not know or care what the underlying type is. I just don't want unused interfaces and unused interface methods in `main` if we can help it.

Reply:
Let me start with: I'm sorry. I was writing this while frustrated for wanting to get this thing landed, knowing that it wasn't going to make it on Friday because of time zones. Everything about the reviews have improved the code and I'm immensely grateful for them.

The point I (unsuccessfully) was trying to present in my previous comment is that I don't think I can (easily) go from c27caff8073868eee9e62a5fce2784badf8de098 (which was very early in this process) to ae13799cc89d2090827b94a896f375da39554f67 (which is nearly a platonic ideal) without using the tools which get removed in 71b515e2a11ebf185eeccd1755942624d61b724b.

The interfaces were the stepping stone; they're indicative of an incomplete migration. The fact that they leak into needing application code to deal with typechecks on them makes them feel gross, and strongly encourages their eventual removal. However, shipping something that is an interface as part of a migration step is still a net improvement since it acts as a forcing function for the author handle the complexity of, "we don't know what type of path we've been handed."

This PR itself has, by virtue of continued investment, evolved beyond needing the interfaces. But that's not necessarily the place where the effort should be stopped on all changes (and arguably, not where this one should have gone to either). The interfaces enable incrementalism. We could go through and tag everything as a `FilePathInterface` tomorrow and within a month we'd have fully-typed paths throughout the codebase because a `FilePathInterface` is such a pain to work with.

I don't know how good Go is at dropping things, but I assume that since it does a SSA pass (that only bails for generics), anything that isn't used should get dropped at compile time. I'd like to revert 71b515e2a11ebf185eeccd1755942624d61b724b, but the other changes were a net positive.

Reply:
I don't want to add unused code to main. Let's cut this and open another PR at another time. PR has been open over a month and we need to ship

Reply:
We can ship path abstraction in 1.3.x